crates/executor/src/blockchain_tree/mod.rs

```diff
@@ -0,0 +1,248 @@
+    fn insert_chain(&mut self, chain: Chain) -> ChainId {
+        let chain_id = self.chain_id_generator;
+        self.chain_id_generator += 1;
+        self.block_indices.insert_chain(chain_id, &chain);
+        // add chain_id -> chain index
+        self.chains.insert(chain_id, chain);
+        chain_id
+    }
+
+    /// DONE
+    /// Insert block inside tree
+    pub fn insert_block(&mut self, block: SealedBlock) -> Result<(), Error> {
+        // check if block number is inside pending block slide
+        if block.number <= self.finalized_block {
+            return Err(ExecError::PendingBlockIsFinalized {
+                block_number: block.number,
+                block_hash: block.hash(),
+                last_finalized: self.finalized_block,
+            }
+            .into())
+        }
+
+        // we will not even try to insert blocks that are too far in future.
+        if block.number > self.finalized_block + self.max_chain_length {
+            return Err(ExecError::PendingBlockIsInFuture {
+                block_number: block.number,
+                block_hash: block.hash(),
+                last_finalized: self.finalized_block,
+            }
+            .into())
+        }
+
+        // check if block parent can be found in Tree
+        if let Some(parent_chain) = self.block_indices.get_block_chain_id(&block.parent_hash) {
+            let _ = self.join_block_to_chain(block.clone(), parent_chain)?;
+            self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
+            return Ok(())
+        }
+
+        // if not found, check if it can be found inside canonical chain.
+        if Some(block.parent_hash) == self.block_indices.canonical_hash(&(block.number - 1)) {
+            // create new chain that points to that block
+            let chain = Chain::new_canonical_joint(&block, &self.db, &self.consensus)?;
+            self.insert_chain(chain);
+            self.db.tx_mut()?.put::<tables::PendingBlocks>(block.hash(), block.unseal())?;
+            return Ok(())
+        }
+        // NOTE: Block dont have parent, and if we receive this block in `make_canonical` function
+        // this could be a trigger to initiate syncing, as we are missing parent.
+        Ok(())
+    }
+
+    // DONE
+    /// Do finalization of blocks. Remove them from tree
+    pub fn finalize_block(&mut self, finalized_block: BlockNumber) {
+        let mut remove_chains = self.block_indices.finalize_canonical_blocks(&finalized_block);
+
+        while let Some(chain_id) = remove_chains.first() {
+            if let Some(chain) = self.chains.remove(chain_id) {
+                remove_chains.extend(self.block_indices.remove_chain(&chain));
+            }
+        }
+        self.finalized_block = finalized_block;
+    }
+
+    /// DONE
+    /// Make block and its parent canonical. Unwind chains to database if necessary.
+    pub fn make_canonical(&mut self, block_hash: &BlockHash) -> Result<(), ()> {
```

Comment:
this is the method that should be called on forkchoice updated, right?

Reply:
Yes, if `BlockchanTree` does not know the hash (aka the block) it will return the error and engine API can switch from the mode from  consensus sync to initial sync to get the blocks from p2p network.

Error needs to be make but the line below should do that.

Reply:
Just to say, if we receive the block in new_payload that we don't know the parent, we are discarding it atm as to not allow potential "bad" blocks inside the Tree. This can be changed in future if neded and the last N blocks can be saved as `unchained` blocks or something.